chain-core/src/init/config.rs

```diff
@@ -21,6 +22,8 @@ pub struct InitConfig {
     pub launch_incentive_to: RedeemAddress,
     // 0x71507ee19cbc0c87ff2b5e05d161efe2aac4ee07 on Eth mainnet ERC20
     pub long_term_incentive: RedeemAddress,
+    // Initial fee setting
+    pub initial_fee_policy: LinearFee,
```

Comment:
This should be:
```
pub struct InitConfig<T> where T: FeeAlgorithm {
    // ... other fields ... //

    pub initial_fee_policy: T,
}
```

Reply:
"Init" is and will only be done once (and affects only a small portion of code), so I'm not sure how it'll help here? For the ABCI app / state, it'll be good to specify it against a trait, but not sure how it helps here?

Reply:
It depends on the intent we want to demonstrate from our design. Current design indicates that `initial_fee_policy` will always be of type `LinearFee` and should not be changed. On the other hand, using generic type bound by `FeeAlgorithm` conveys an intent that `initial_fee_policy` can be changed in future but should always implement `FeeAlgorithm` trait. It doesn't matter how many times a function will be called at runtime (all the generic type information will anyways be erased at runtime).

Even though the amount of code that gets affected by changing `initial_fee_policy` is currently very small (this can grow in future), we should change it if our intent is to make it easily switchable.

Reply:
added a TODO comment there, as the scope of changing this is quite large

Reply:
but not sure if `InitConfig` should be easily switchable -- genesis is fixed... so once the network starts, this is frozen. The current design's intent captures that.
For `ChainNodeState`, it's different -- that one may change, so it'll be worth to specify it against a trait.

Reply:
I'll reiterate what I said earlier. Generics do not give you ability to change type of an object at runtime. For example, you cannot do this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1e759e63c2d1c16ca8659f8e24560ab5.

Generics are only to help developers to make changes in implementation without affecting large parts of code by hinting the compiler about possible concrete types. Compiler will infer or calculate concrete types at compile time which cannot be changed later during runtime.

So, in our case, even if we make `InitConfig` generic, we wont be able to change it after genesis. It'll just give us _plug-and-play_-ability on `FeeAlgorithm` for easy development.

Reply:
I understand that and my concern isn't about changing a type at runtime -- the point being here is that _plug-and-play_-ability may not be desired in InitConfig, as it could become error prone down the line (if one generic type is simply sprinkled everywhere).
To make it less error-prone, one consequence would be that the code should carry two generic params: `A: FeeAlgorithm + ...`, `B: FeeAlgorithm + ...` where A is `InitConfig` and B is for `ChainNodeState`. Or make sure that InitConfig's generic type is instantiated with a specific type in critical places with some warning comment not to change it.

For `ChainNodeState`, I'm 100% for making it generic.
For `InitConfig`, I'm fine with making generic, but more care would need to be taken and the costs against keeping it monomorphic need to be weighted.

Reply:
Let's start with making `ChainNodeState` generic. We'll make `InitConfig` generic later if we feel it becoming necessary. 
I'm ok with this approach. We just have to keep in mind that doing this change later will take a lot more effort than it'll take now.